File: core/src/main/java/com/codebasecensus/core/CodebaseCensus.java
package com.codebasecensus.core;

import android.os.Environment;
import org.eclipse.jgit.api.Git;
import org.eclipse.jgit.api.TransportConfigCallback;
import org.eclipse.jgit.api.errors.GitAPIException;
import org.eclipse.jgit.transport.Transport;

import java.io.File;
import java.net.URI;
import java.util.Arrays;

/**
 * Created by kevin on 08/12/18 at 11:31
 */
public class CodebaseCensus implements Codebase{
    /**
     *
     * @param uri e.g "https://github.com/eclipse/jgit.git"
     * @param callback
     */
    @Override
    public void cloneRepo(URI uri, CloneCallback callback) {
        cloneRepo(uri, "master",callback);
    }

    @Override
    public void cloneRepo(URI uri, String branch, CloneCallback callback) {
        try {
            Git.cloneRepository()
                    .setURI(uri.toASCIIString())
                    .setDirectory(new File(getRepoDirectory(uri)))
                    .setBranchesToClone(Arrays.asList("refs/heads/"+branch))
                    .setBranch("refs/heads/"+branch)
                    .call();
        } catch (GitAPIException e) {
            e.printStackTrace();
        }
    }

    String getRepoDirectory(URI uri){
        return Environment.getExternalStorageDirectory().getPath()+"/codebase-census/"+folderNameFromGiturl(uri);
    }

    String folderNameFromGiturl(URI uri){
        String url = uri.toASCIIString();

        String[] exploded = url.split("/");

        return exploded[exploded.length-1];
    }
}

File: core/src/main/java/com/codebasecensus/core/CodebaseCensus.kt
package com.codebasecensus.core

import java.io.File
import org.eclipse.jgit.storage.file.FileRepositoryBuilder



/**
 * Created by kevin on 08/12/18 at 14:01
 */
class CodebaseCensus(var git: GIT) : Codebase {

    var initialized : Boolean = false

    override fun setup(callback: SetupCallback) {
        if(git.repoExists()){
            git.pull(object: PullCallback{
                override fun pullSuccess() {
                    initialized = true
                    callback.setupSuccess()
                }

                override fun pullFailed(error: String) {
                    callback.setupFailed(error)
                }
            })
        }else{
            git.cloneRepo("master",object: CloneCallback{
                override fun cloneSuccess() {
                    initialized = true
                    callback.setupSuccess()
                }

                override fun cloneFailure(reason: String) {
                    callback.setupFailed(reason)
                }
            })
        }
    }

    override fun contributors() {

    }

    override fun analizeTotalFileCount(callback: AnalyzeCallback) {
        if(!initialized){
            callback.onFailure("Not initialized, call setup() first")
            return
        }

        val dirString = git.getRepoDirectory()

        Thread(Runnable {
            val count : Int = countFiles(File(dirString))
            val model = ResultModel(AnalysisType.FILE_COUNT,count)
            callback.onFinished(model)
        }).start()
    }

    private fun countFiles(file : File) : Int{
        var size = 0
        for (file in file.listFiles()!!) {
            size += if (file.isFile()) {
                1
            } else {
                countFiles(file)
            }
        }
        return size
    }
}